scipy/linalg/_solvers.py

```diff
@@ -416,23 +441,66 @@ def solve_discrete_are(a, b, q, r):
             raise ValueError("Matrix {} should be symmetric/hermitian."
                              "".format("qr"[ind]))
 
+    # Check if the generalized case is required with omitted arguments
+    # perform late shape checking etc. 
+    generalized_case = e is not None or s is not None
+    e_is_I_or_None = True
```

Comment:
this variable is not needed. only used once, so simply ``if e is not None:`` seems clearer